### ISSUE TITLE
cluster_url with a path is omitted

### ISSUE THREAD

**Issue by ktarplee** [clux/kube-rs#244]
When I make any request, 404 error is thrown.

Using a debugger I was able to track this issue down where this is happening.  My cluster_url includes more than the hostname (i.e. https://hostname/foo/bar) because I am using rancher.  This presents no problems for kubectl.

mod.rs:75
```rust
let uri_str = self.cluster_url.join(pandq.as_str())?;
```

pandq is "/api/v1/nodes?".  The problem is that the resultant uri_str is "https://hostname/api/v1/nodes?" and not what it should be, namely "https://hostname/foo/bar/api/v1/nodes?"

I tested version 0.34.0.

**Comment by ktarplee** [comment]
One possible fix it to use relative paths for "pandq" above (i.e. no preceding slash) and ensure that the cluster_url has a trailing slash.  Then the join will work if cluster_url has more than just a hostname.

See the documentation for [url:Url.join()](https://docs.rs/url/2.1.1/url/struct.Url.html#method.join)

**Comment by ktarplee** [comment]
I should mention that this is a regression since 0.24.0 (i.e., it worked fine in 0.24.0 because the code you used was a string join instead of the uri.join).

**Comment by clux** [comment]
Yeah, this if definitely a bug, thanks you!

Even remember doing the url join at that time and being frustrated with the docs at the time.

**Comment by clux** [comment]
Presumably this would work if the `cluster_url` specified in your ~/.kube/config had a trailing slash?
We could work with this behaviour of Url::join by always suffixing one in our config module if I understand it correctly.

The intention of the code basically looks correct here, but it's just a weird edge case in how `.join` works.

**Comment by clux** [comment]
Introduced in #208

**Comment by ktarplee** [comment]
@clux You also need to remove the slash at the front of the pandq url (or any URL joined to the cluster_url.  In other words you have to make the URL you are joining to be a relative URL and not absolute.  If it is absolute, join() will do the right thing and remove all but the path from cluster_url.

**Comment by ktarplee** [comment]
I'm not sure this is the only place it occurs but you need to remove the slash in front of "group"

```rust
pub(crate) fn make_url(&self) -> String {
        let n = if let Some(ns) = &self.namespace {
            format!("namespaces/{}/", ns)
        } else {
            "".into()
        };
        format!(
            "/{group}/{api_version}/{namespaces}{resource}",
            group = if self.group.is_empty() { "api" } else { "apis" },
            api_version = self.api_version,
            namespaces = n,
            resource = to_plural(&self.kind.to_ascii_lowercase()),
        )
    }
```

**Comment by clux** [comment]
well, i can take out the leading slash in make_url (which is really creating the path part of the url), but then that's not really in line with any docs: Url is talking about the trailing slash they mention as having special behaviour, and our function is well-defined with a leading slash, and is also pretty heavily unit tested as such.

ideally we need to take care of both path segments and combine them carefully:
- the path segment coming from rancher's hostname
- the path coming from kube's api module

we have guarantees from the api module that we always have the leading slash, and it's only the cluster_url which is variable user input.

**Comment by clux** [comment]
Convinced this needs to go into the client module with this an interface that takes a `url::Url` from config + a [http::Uri](https://docs.rs/http/0.2.1/http/uri/struct.Uri.html) from the http request, and creates a combo string.

A failing test I wrote:

```rust
    #[test]
    fn rancher_host_success() {
        let rancher_host = "https://hostname/foo/bar";
        let cluster_url = reqwest::Url::parse(rancher_host).unwrap();
        let apipath = Url::parse("/api/v1/nodes?").unwrap();

        //let final_url = ???
        assert_eq!(final_url.as_str(), "https://hostname/foo/bar/api/v1/nodes");
    }
```

**Comment by clux** [comment]
Ok. Have a solution that passes both normal tests. If you could check that would be great.

**Comment by clux** [comment]
Fix released in 0.35.0
